File: problems/arrays/count_squares/python/count_squares.py
def countSquares(points: list[list[int]]) -> int:
    point_map = {tuple(point): True for point in points}
    count = 0
    rotation_clockwise = [0, 1, -1, 0]
    rotation_counter_clockwise = [0, -1, 1, 0]

    def rotate(point: dict, matrix: list[int]) -> dict:
        return {
            "x": matrix[0] * point["x"] + matrix[1] * point["y"],
            "y": matrix[2] * point["x"] + matrix[3] * point["y"],
        }

    def has_point(point: dict) -> bool:
        return tuple([point["x"], point["y"]]) in point_map

    def check_square(p1: dict, p2: dict, vector: dict, matrix: list[int]) -> int:
        rotated_vector = rotate(vector, matrix)
        p3 = {"x": p1["x"] + rotated_vector["x"], "y": p1["y"] + rotated_vector["y"]}
        p4 = {"x": p2["x"] + rotated_vector["x"], "y": p2["y"] + rotated_vector["y"]}
        return 1 if has_point(p3) and has_point(p4) else 0

    for i in range(len(points)):
        p1 = {"x": points[i][0], "y": points[i][1]}
        for j in range(i + 1, len(points)):
            p2 = {"x": points[j][0], "y": points[j][1]}

            dx = p2["x"] - p1["x"]
            dy = p2["y"] - p1["y"]

            count += check_square(p1, p2, {"x": dx, "y": dy}, rotation_clockwise)
            count += check_square(
                p1, p2, {"x": dx, "y": dy}, rotation_counter_clockwise
            )

    return count // 4

File: problems/arrays/count_squares/js/count_squares.js
export const countSquares = (points) => {
  const pointMap = new Map();
  for (const [x, y] of points) pointMap.set([x, y].toString(), true);

  let count = 0;
  const rotationClockwise = [0, 1, -1, 0];
  const rotationCounterClockwise = [0, -1, 1, 0];

  const rotate = (point, matrix) => {
    return {
      x: matrix[0] * point.x + matrix[1] * point.y,
      y: matrix[2] * point.x + matrix[3] * point.y,
    };
  };

  const hasPoint = (point) => pointMap.has(point.toString());

  const checkSquare = (p1, p2, vector, matrix) => {
    const rotatedVector = rotate(vector, matrix);
    const p3 = [p1.x + rotatedVector.x, p1.y + rotatedVector.y];
    const p4 = [p2.x + rotatedVector.x, p2.y + rotatedVector.y];

    return hasPoint(p3) && hasPoint(p4) ? 1 : 0;
  };

  for (let i = 0; i < points.length; i++) {
    const p1 = { x: points[i][0], y: points[i][1] };
    for (let j = i + 1; j < points.length; j++) {
      const p2 = { x: points[j][0], y: points[j][1] };

      const dx = p2.x - p1.x;
      const dy = p2.y - p1.y;

      count += checkSquare(p1, p2, { x: dx, y: dy }, rotationClockwise);
      count += checkSquare(p1, p2, { x: dx, y: dy }, rotationCounterClockwise);
    }
  }
  return count / 4;
};
